core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -16,60 +16,26 @@
 
 package scalaz.zio
 
-import scala.annotation.tailrec
 import scalaz.zio.internal.Platform
-import scalaz.zio.Queue.internal._
+import scalaz.zio.Queue2.internal._
 import scalaz.zio.internal.MutableConcurrentQueue
 
+import scala.annotation.tailrec
+
 /**
- *  A `Queue[A]` is a lightweight, asynchronous queue for values of type `A`.
+ * A `Queue2[R, E, A, B]` is a lightweight, asynchronous queue into which values of type `A`
+ * can be enqueued and of which elements of type `B` can be dequeued. The queue's enqueueing
+ * operations may utilize an environment of type `RA` and may fail with errors of type `EA`.
+ * The dequeueing operations may utilize an environment of type `RB` and may fail with errors
+ * of type `EB`.
  */
-class Queue[A] private (
-  queue: MutableConcurrentQueue[A],
-  takers: MutableConcurrentQueue[Promise[Nothing, A]],
-  shutdownHook: Ref[Option[UIO[Unit]]],
-  strategy: Strategy[A]
-) extends Serializable {
-
-  private final val checkShutdownState: UIO[Unit] =
-    shutdownHook.get.flatMap(_.fold[UIO[Unit]](IO.interrupt)(_ => IO.unit))
-
-  @tailrec
-  private final def pollTakersThenQueue(): Option[(Promise[Nothing, A], A)] =
-    // check if there is both a taker and an item in the queue, starting by the taker
-    if (!queue.isEmpty()) {
-      val nullTaker = null.asInstanceOf[Promise[Nothing, A]]
-      val taker     = takers.poll(nullTaker)
-      if (taker == nullTaker) {
-        None
-      } else {
-        queue.poll(null.asInstanceOf[A]) match {
-          case null =>
-            unsafeOfferAll(takers, taker :: unsafePollAll(takers))
-            pollTakersThenQueue()
-          case a => Some((taker, a))
-        }
-      }
-    } else None
-
-  @tailrec
-  private final def unsafeCompleteTakers(platform: Platform): Unit =
-    pollTakersThenQueue() match {
-      case None =>
-      case Some((p, a)) =>
-        unsafeCompletePromise(p, a, platform)
-        strategy.unsafeOnQueueEmptySpace(queue, platform)
-        unsafeCompleteTakers(platform)
-    }
-
-  private final def removeTaker(taker: Promise[Nothing, A]): UIO[Unit] = IO.effectTotal(unsafeRemove(takers, taker))
-
-  final val capacity: Int = queue.capacity
+trait Queue2[-RA, +EA, -A, -RB, +EB, +B] extends Serializable { self =>
```

Comment:
Even though types are not curried in Scala, how do you feel about keeping `A` and `B` in tail position?

Reply:
Good point, otherwise typeclass instances aren't possible when integrating with cats/scalaz.